[GTK][WPE][GStreamer] Fix build with glib < 2.68

magomez opened the pull request:
#### e6024502b5ca9beca469edbd615db07c5fc017a5
<pre>
[GTK][WPE][GStreamer] Fix build with glib &lt; 2.68
<a href="https://bugs.webkit.org/show_bug.cgi?id=279177">https://bugs.webkit.org/show_bug.cgi?id=279177</a>

Reviewed by Philippe Normand and Michael Catanzaro.

Replace the usage of g_memdup2 with an equivalent code.

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(gst_buffer_new_memdup):

Canonical link: <a href="https://commits.webkit.org/283216@main">https://commits.webkit.org/283216@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/449fad69ed7c6d250bf4084603669f4b2c5eabec

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65578 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44949 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18196 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69604 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16187 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67696 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52750 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16467 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52657 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11231 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68645 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41520 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56760 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33285 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38200 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14139 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15063 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60033 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14480 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71310 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9532 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13926 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59978 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9564 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56825 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60253 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7878 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1525 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9937 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40759 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41835 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43018 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41579 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -1557,7 +1557,11 @@ std::optional<unsigned> gstGetAutoplugSelectResult(ASCIILiteral nick)
+#if GLIB_CHECK_VERSION(2, 68, 0)
     gpointer copiedData = g_memdup2(data, size);
+#else
+    gpointer copiedData = g_memdup(data, size);
```
This is an integer truncation vulnerability (which is why `g_memdup()` is deprecated). Now if `size > UINT_MAX` then GstBuffer thinks it has a larger buffer than it really does.

I would just open code `g_malloc()` then `memcpy()`. Or you could truncate safely by doing `size = static_cast<unsigned>(size)`.

magomez replied:
let's do that truncation then.

mcatanzaro commented on Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
```diff
@@ -1557,7 +1557,11 @@ std::optional<unsigned> gstGetAutoplugSelectResult(ASCIILiteral nick)
+    gpointer copiedData = g_memdup(data, static_cast<unsigned>(size));
+#endif
 
     return gst_buffer_new_wrapped_full(static_cast<GstMemoryFlags>(0), copiedData, size, 0, size, copiedData, g_free);
```
The cast would need to be here:

```suggestion
    return gst_buffer_new_wrapped_full(static_cast<GstMemoryFlags>(0), copiedData, size, 0, static_cast<unsigned>(size), copiedData, g_free);
```

Though of course you don't want to do that in the case where you do have `g_memdup2()`.

Yet another option: you could copy the entire function:

```
gpointer
g_memdup2 (gconstpointer mem,
           gsize         byte_size)
{
  gpointer new_mem;

  if (mem && byte_size != 0)
    {
      new_mem = g_malloc (byte_size);
      memcpy (new_mem, mem, byte_size);
    }
  else
    new_mem = NULL;

  return new_mem;
}
```

magomez replied:
> The cast would need to be here:

Ah, ok, I get what you mean. The cast that I added is not needed because there's already an implicit cast to guint happening. But when calling g_memdup we may get an amount of memory smaller than it was requested because of the cast. This means that the behavior won't be exactly the same in both cases. So I think the proper fix is copying the function, as you mention.

magomez committed:
[GTK][WPE][GStreamer] Fix build with glib < 2.68
https://bugs.webkit.org/show_bug.cgi?id=279177

Reviewed by Philippe Normand and Michael Catanzaro.

Replace the usage of g_memdup2 with an equivalent code.

* Source/WebCore/platform/graphics/gstreamer/GStreamerCommon.cpp:
(gst_buffer_new_memdup):

Canonical link: https://commits.webkit.org/283216@main
